test/jdk/java/awt/List/ListGarbageCollectionTest/AwtListGarbageCollectionTest.java

```diff
@@ -44,30 +48,36 @@ public static void main(String[] args) {
         }
     }
 
-    private static void test(Frame frame) {
-        WeakReference<List> weakListRef = null;
-        try {
-            frame.setSize(300, 200);
-            frame.setVisible(true);
+    private static void test(Frame frame) throws InterruptedException {
+        frame.setSize(300, 200);
+        frame.setVisible(true);
+
+        List strongListRef = new List();
+        frame.add(strongListRef);
+        strongListRef.setMultipleMode(true);
+        frame.remove(strongListRef);
+
+        final ReferenceQueue<List> referenceQueue = new ReferenceQueue<>();
+        final PhantomReference<List> phantomListRef =
+                new PhantomReference<>(strongListRef, referenceQueue);
+        System.out.println("phantomListRef: " + phantomListRef);
+
+        strongListRef = null; // Clear the strong reference
 
-            List strongListRef = new List();
-            frame.add(strongListRef);
-            strongListRef.setMultipleMode(true);
-            frame.remove(strongListRef);
-            weakListRef = new WeakReference<List>(strongListRef);
-            strongListRef = null;
+        System.out.println("Waiting for the reference to be cleared");
+        int count = 1;
+        Reference<? extends List> ref;
+        do {
+            System.out.println("Attempt " + count);
+            System.gc();
```

Comment:
As far as I remember the call of System.gc() could be ignored, and the only way to trigger GC is to trigger the OOM. Probably we can combine referenceQueue.remove(ENQUEUE_TIMEOUT) + OOM?

Reply:
For this reason, GC logs are enabled. If, for whatever reason, the call to `System.gc()` is ignored, we'll see it in the test log. Then a specific GC could be selected, for example, or another fix implemented.

OOME does not guarantee a full GC cycle either.

Having the above in mind, I'd rather keep it simple.

Reply:
>For this reason, GC logs are enabled. If, for whatever reason, the call to System.gc() is ignored, we'll see it in the test log. Then a specific GC could be selected, for example, or another fix implemented.

As of now this test can be executed with different GC, and some of them can skip System.gc().

>OOME does not guarantee a full GC cycle either.

Reply:
>For this reason, GC logs are enabled. If, for whatever reason, the call to System.gc() is ignored, we'll see it in the test log. Then a specific GC could be selected, for example, or another fix implemented.

As of now this test can be executed with different GC, and some of them can skip System.gc().

>OOME does not guarantee a full GC cycle either.

But it guarantee that at list some GC will be always run, unlike System.gc().

Reply:
I followed the piece of advice in the [JBS comment](https://bugs.openjdk.org/browse/JDK-8300727?focusedCommentId=14560139&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14560139).

If using `System.gc()` is good enough for testing references:

https://github.com/openjdk/jdk/blob/f612dcfebea7ffd4390f833646ad45d6f0ebd04f/test/jdk/java/lang/ref/ReferenceEnqueue.java#L54-L60

https://github.com/openjdk/jdk/blob/f612dcfebea7ffd4390f833646ad45d6f0ebd04f/test/jdk/java/lang/ref/PhantomReferentClearing.java#L85-L92

Then it should be good enough for this test too.

Perhaps, the same effect could be achieved by causing OOME in a loop. However, using `System.gc()` makes the intention clearer: *run GC*, wait for the phantom reference to be cleared and enqueued.

I can explicitly select a GC: `-XX:+UseG1GC` (default) or `-XX:+UseSerialGC`.

Reply:
There are lots of tests that rely on System.gc() actually triggering a gc.
SFAIK all collectors we have today will obey it unless you use
-XX:+DisableExplicitGC
If you can make the test work with that then that would be interesting but calling just System.gc() is no worse than all those other tests.

Reply:
Many client tests generate OOM for that, we have a special Util, see the usage of Util.generateOOME() or various implementation of generateOOME.

If the OOM + sleep(to give the GC a chance to clean the weakrefs in case of slow systems) does not work, then could it be considerred as a GC bug?

>OutOfMemoryError: The Java Virtual Machine implementation has run out of either virtual or physical memory, and the automatic storage manager was unable to reclaim enough memory to satisfy an object creation request.
https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-6.html#jvms-6.3

Reply:
Yes, I saw this method, and [it periodically **calls `System.gc()`**](https://github.com/openjdk/jdk/blob/b4ea80731c6c0a0328a9801590ba5b081f08c3bd/test/jdk/javax/swing/regtesthelpers/Util.java#L102-L103).

I still don't understand your concern. How is generating `OutOfMemoryError` better than calling `System.gc()`?

Calling `System.gc()` conveys the intent in a clearer way. No OOME is required.